drand/drand_test.go

```diff
@@ -22,3 +27,159 @@ func TestNew(test *testing.T) {
+	host, err := p2pimpl.NewHost(&leader, priKey)
+	if err != nil {
+		test.Fatalf("newhost failure: %v", err)
+	}
+	dRand := New(host, "0", []p2p.Peer{leader, validator}, leader, nil, true)
+
+	if !dRand.IsLeader {
+		test.Error("dRand should belong to a leader")
+	}
+
+	newPeer := p2p.Peer{IP: "127.0.0.1", Port: "9907"}
+	countValidatorPeers := dRand.AddPeers([]*p2p.Peer{&newPeer})
+
+	if countValidatorPeers != 1 {
+		test.Error("Unable to add new peer")
+	}
+}
+
+func TestGetValidatorByPeerId(test *testing.T) {
+	leader := p2p.Peer{IP: "127.0.0.1", Port: "9902"}
+	validator := p2p.Peer{IP: "127.0.0.1", Port: "9905"}
+	priKey, _, _ := utils.GenKeyP2P("127.0.0.1", "9902")
+	host, err := p2pimpl.NewHost(&leader, priKey)
+	if err != nil {
+		test.Fatalf("newhost failure: %v", err)
+	}
+	dRand := New(host, "0", []p2p.Peer{leader, validator}, leader, nil, true)
+
+	if !dRand.IsLeader {
+		test.Error("dRand should belong to a leader")
+	}
+
+	validatorID := utils.GetUniqueIDFromPeer(validator)
+
+	if dRand.getValidatorPeerByID(validatorID) == nil {
+		test.Error("Unable to get validator by Peerid")
+	}
+	newPeer := p2p.Peer{IP: "127.0.0.1", Port: "9907"}
+	newPeerID := utils.GetUniqueIDFromPeer(newPeer)
+
+	if dRand.getValidatorPeerByID(newPeerID) != nil {
+		test.Error("Found validator for absent validatorId")
+	}
+}
+
+func TestResetState(test *testing.T) {
+	leader := p2p.Peer{IP: "127.0.0.1", Port: "9902"}
+	validator := p2p.Peer{IP: "127.0.0.1", Port: "9905"}
+	priKey, _, _ := utils.GenKeyP2P("127.0.0.1", "9902")
+	host, err := p2pimpl.NewHost(&leader, priKey)
+	if err != nil {
+		test.Fatalf("newhost failure: %v", err)
+	}
+	dRand := New(host, "0", []p2p.Peer{leader, validator}, leader, nil, true)
+	dRand.ResetState()
+}
+
+func TestSetLeaderPubKey(test *testing.T) {
+	leader := p2p.Peer{IP: "127.0.0.1", Port: "9902"}
+	validator := p2p.Peer{IP: "127.0.0.1", Port: "9905"}
+	priKey, _, _ := utils.GenKeyP2P("127.0.0.1", "9902")
+	host, err := p2pimpl.NewHost(&leader, priKey)
+	if err != nil {
+		test.Fatalf("newhost failure: %v", err)
+	}
+	dRand := New(host, "0", []p2p.Peer{leader, validator}, leader, nil, true)
+
+	_, newPublicKey, _ := utils.GenKeyP2P("127.0.0.1", "9902")
+	newPublicKeyBytes, _ := newPublicKey.Bytes()
+	if dRand.SetLeaderPubKey(newPublicKeyBytes) == nil {
+		test.Error("Failed to Set Public Key of Leader")
+	}
+}
+
+func TestUpdatePublicKeys(test *testing.T) {
+	leader := p2p.Peer{IP: "127.0.0.1", Port: "9902"}
+	validator := p2p.Peer{IP: "127.0.0.1", Port: "9905"}
+	priKey, _, _ := utils.GenKeyP2P("127.0.0.1", "9902")
+	host, err := p2pimpl.NewHost(&leader, priKey)
+	if err != nil {
+		test.Fatalf("newhost failure: %v", err)
+	}
+	dRand := New(host, "0", []p2p.Peer{leader, validator}, leader, nil, true)
+
+	_, pubKey1 := utils.GenKey("127.0.0.1", "5555")
+	_, pubKey2 := utils.GenKey("127.0.0.1", "6666")
+
+	if dRand.UpdatePublicKeys([]*bls.PublicKey{pubKey1, pubKey2}) != 2 {
```

Comment:
any better way to make make the public keys are updated properly?

Reply:
Along with the count, I am comparing the hexString of publicKeys which are updated.